src/erlfmt_ast.erl

```diff
@@ -0,0 +1,1086 @@
+                        T = erl_syntax:class_qualifier_stacktrace(Pat),
+                        Ps =
+                            case erl_syntax:type(T) of
+                                underscore -> [A, B];
+                                _ -> [A, B, T]
+                            end,
+                        %% have to figure out start and end pos for catch node
+                        CPos = erl_syntax:get_pos(Clause),
+                        PPos = until(right_pos(Ps, CPos), 0, CPos),
+                        {'catch', PPos, [st_to_erlfmt(P) || P <- Ps]};
+                    _ ->
+                        st_to_erlfmt(Pat)
+                end;
+            {name, Name0} ->
+                %% must copy the line from the individual clause
+                %% onto the name node
+                Name = set_line_from(Clause, Name0),
+                Patterns = erl_syntax:clause_patterns(Clause),
+                %% have to figure out start and end pos for call node
+                NPos = erl_syntax:get_pos(Name),
+                % add 1 for start paren
+                NPos1 = until(NPos, 1, NPos),
+                APos = until(right_pos(Patterns, NPos1), 1, NPos),
+                st_to_erlfmt(
+                    erl_syntax:set_pos(
+                        erl_syntax:application(Name, Patterns),
+                        APos
+                    )
+                );
+            'fun' ->
+                Patterns = erl_syntax:clause_patterns(Clause),
+                %% have to figure out start and end pos for args node
+                CPos = erl_syntax:get_pos(Clause),
+                % 1 for start paren
+                CPos1 = abswidth(1, CPos),
+                APos = until(right_pos(Patterns, CPos1), 1, CPos),
+                {args, APos, [st_to_erlfmt(P) || P <- Patterns]}
+        end,
+    Guard =
+        case erl_syntax:clause_guard(Clause) of
+            none -> empty;
+            G -> st_guard_to_erlfmt(G)
+        end,
+    Body = [st_to_erlfmt(E) || E <- erl_syntax:clause_body(Clause)],
+    {clause, Pos, Head, Guard, Body}.
+
+st_guard_to_erlfmt(Guard) ->
+    Pos = erl_syntax:get_pos(Guard),
+    case erl_syntax:type(Guard) of
+        conjunction ->
+            {guard_and, Pos, [
+                st_guard_to_erlfmt(G)
+                || G <- erl_syntax:conjunction_body(Guard)
+            ]};
+        disjunction ->
+            {guard_or, Pos, [
+                st_guard_to_erlfmt(G)
+                || G <- erl_syntax:disjunction_body(Guard)
+            ]};
+        _ ->
+            st_to_erlfmt(Guard)
+    end.
+
+st_typedecl_to_erlfmt({atom, _, type} = Tag, [Term], Pos) ->
+    %% the argument is raw erl_parse data here, lifted to abstract form by
+    %% erl_syntax, so it first needs to be made concrete again
+    {Name, Type, Args} = erl_syntax:concrete(Term),
+    Name1 = {atom, dummy_anno(Name), Name},
+    Type1 = st_to_erlfmt(Type),
+    %exit({Name1, Type, Type1}),
```

Comment:
What is this comment for?

Reply:
Because I'm daft and forgot to remove it. It was there to provoke a visible crash at some point while debugging.

Reply:
fixed